keps/sig-cloud-provider/20190423-physical-host-labelling.md

```diff
@@ -0,0 +1,268 @@
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Many kubernetes clusters are deployed on virtual machines (also referred to as
+"instances"), and information about the physical host for those instances (also
+referred to as "instance host") is not consistently available. Instances which
+share a host can fail simultaneously if there is a problem with the host. We
+propose to label nodes with their instance host ID so that controllers (namely,
+the scheduler) can decrease shared fate of pods from the same collection.
+
+## Motivation
+
+This KEP is motivated by the desire to improve reliability of workloads running
+in kubernetes clusters. Today, the scheduler will attempt to spread pods from a
+given collection (e.g. a deployment) across nodes. Even if every pod ends up on
+a different node, there may be multiple pods per instance host. This means that
+the deployment is not as reliable as it appears to be.
+
+By adding a physical host label, we can treat physical hosts as a failure
+domain, and perform spreading appropriately.
+
+### Goals
+
+- Introduce a well known label for physical host (e.g.
+  `topology.kubernetes.io/physical-host`)
+- Add a cloud provider API to get physical host ID
+- Update cloud-controller-manager to update node labels based on the cloud
+  provider value.
+
+### Non-Goals
+
+- Scheduling behavior that considers physical host
+- Controller behavior (e.g. deployment, statefulset) that considers physical
+  host
+- Any decision-making that is based on the physical host label
+
+## Proposal
+
+We will add a new field to the [Zone struct][zone-struct], `PhysicalMachineID`.
+If supported, it will be populated by cloud providers whenever returning a zone.
+
+[zone-struct]: https://github.com/kubernetes/cloud-provider/blob/0a4f4cbb5a664deb4639d7d9bf5bbde3bb3603c1/cloud.go#L208-L211
+
+When the `PhysicalMachineID` is supplied, cloud-controller-manager will use it
+as the value for the `topology.kubernetes.io/physical-host` label. (Depending on
+when this is implemented, this may instead be
+`failure-domain.beta.kubernetes.io/physical-host`. See
+https://github.com/kubernetes/enhancements/pull/839)
+
+### Implementation Details/Notes/Constraints [optional]
+
+What are the caveats to the implementation?
```

Comment:
Here we should list the challenges of what I used to call "label drift" (see [1] for details), i.e. what happens in clusters where VM placement is dynamic. 

Take for example VMware vSphere (applies to other hypervisors as well, incl. Google Cloud with [Live Migration](https://cloud.google.com/compute/docs/instances/live-migration) I suppose) where various events (manual, host maintenance, failure) can lead to a VM to be relocated. In case of HA (reboot), the node would possible re-lable itself after boot, but usually administrators (or cluster rebalancing) thrive to maintain the desired state (e.g. vSphere DRS affinity/anti-affinity rules), leading to live migration (vMotion) of a VM back to a compliant host. The behavior when/if this happens is non-deterministic which complicates the matter.

**tl;dr:** advanced hypervisor features like vSphere DRS/vMotion et al, used in almost every on-prem customer environment for improved cluster utilization, fairness and availability, lead to `topology.kubernetes.io/physical-host` label drift. According to a discussion on SIG Scheduling, the scheduler code might work fine here (cache invalidation/update with new labels <- kind of a relabeler needed to change this node label), but that could invalidate former scheduling decisions (e.g. affinity/anti-affinity).

[1]: [vSphere Cloud Provider should support implement Zones() interface #64021
](https://github.com/kubernetes/kubernetes/issues/64021)

Reply:
Agreed - this area needs careful consideration. Do you think that should be mentioned in the summary/motivation section, or in the follow up PR where we start fleshing out the design?

Reply:
I think this warrants discussion now because it may determine the validity of the topology label in the first place

Reply:
+1 on what @andrewsykim said

Reply:
Generally speaking, Kubernetes scheduling decisions are valid at the moment they are made. There is no guarantee that they remain valid in the future. For example, Pod B is scheduled on the the same node that runs Pod A when Pod B has inter-pod affinity to Pod A, but a second later Pod A may terminate. Similarly, node labels may be updated/deleted. These changes may invalidate NodeAffinity and NodeSelector constraints or pods.
We have accepted the fact that clusters are dynamic in nature and scheduling decisions may be invalidated in the future. This is an assumption that other new features should make too.
We may one day add "Descheduler" as a standard component to find such invalidations and deschedule pods. Whether we add a descheduler or not, IMO the fact that dynamic cluster changes may invalidate a scheduling decision should not be a blocker for new features.